### PR TITLE
feat: adjust DiskStore compact to only store cached data

### PR DESCRIPTION
**Pull request by cryptonemo**
feat: adjust LevelCacheStore for the above
feat: add external reader (callback) option for LevelCacheStore

### CODE REVIEW

File: src/store.rs [hunk 1]
```diff
@@ -610,18 +628,34 @@ pub struct LevelCacheStore<E: Element> {
...
+    // layer data.
+    reader: Option<ExternalReader<R>>,
+
     _e: PhantomData<E>,
 }
 
-impl<E: Element> ops::Deref for LevelCacheStore<E> {
+impl<E: Element, R: Read + Send + Sync> fmt::Debug for LevelCacheStore<E, R> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "LevelCacheStore")
```
Comment (dignifiedquire) [inline src/store.rs]: should write some details of its fields here, see https://doc.rust-lang.org/std/fmt/struct.DebugStruct.html for an easy way to do this

File: src/store.rs [hunk 2]
```diff
@@ -610,18 +628,34 @@ pub struct LevelCacheStore<E: Element> {
...
+    }
+}
+
+impl<E: Element, R: Read + Send + Sync> LevelCacheStore<E, R> {
+    pub fn set_external_reader(&mut self, reader: ExternalReader<R>) {
```
Comment (dignifiedquire) [inline src/store.rs]: as dicussed, lets move this to the constructor instead of a method

File: src/store.rs [hunk 3]
```diff
@@ -824,10 +859,25 @@ impl<E: Element> LevelCacheStore<E> {
...
+            self.reader
+                .as_ref()
+                .unwrap()
+                .read(start, end, &mut read_data)
+                .unwrap_or_else(|_| {
+                    panic!(
+                        "failed to read {} bytes externally at offset {}",
+                        read_len, start
+                    )
+                });
```
Comment (dignifiedquire) [inline src/store.rs]: this should return an error, not panic

File: src/test_xor128.rs [hunk 4]
```diff
@@ -346,6 +350,121 @@ fn test_large_tree() {
...
+            .unwrap();
+
+        let output_file = temp_path.join(format!("base-data-only-{}", i));
+        let mut writer = OpenOptions::new()
+            .write(true)
+            .create(true)
+            .open(&output_file)
+            .context("Failed to open output file")
+            .unwrap();
+        writer
```
Comment (dignifiedquire) [inline src/test_xor128.rs]: you can just use `std::fs::write`

File: src/test_xor128.rs [hunk 5]
```diff
@@ -346,6 +350,121 @@ fn test_large_tree() {
...
+        let reader = OpenOptions::new()
+            .read(true)
+            .open(&output_file)
+            .context("Failed to open base layer data")
+            .unwrap();
+
+        // Compact the disk store for use as a LevelCacheStore.
+        match mt_disk.compact(config.clone()) {
+            Ok(x) => assert_eq!(x, true),
+            Err(_) => continue, // Could not do any compaction with this configuration.
```
Comment (dignifiedquire) [inline src/test_xor128.rs]: shouldn't this panic in the test, such that we are sure this actually happens?

File: src/test_xor128.rs [hunk 6]
```diff
@@ -346,6 +350,121 @@ fn test_large_tree() {
...
+        }
+
+        // Then re-create an MT using LevelCacheStore and generate all proofs.
+        let mut level_cache_store: LevelCacheStore<[u8; 16], _> =
+            Store::new_from_disk(2 * count - 1, &config).unwrap();
+        level_cache_store.set_external_reader(ExternalReader {
+            source: reader,
+            read_fn: |start, end, buf, reader| {
+                reader.read_at(&mut buf[0..end - start], start as u64)
```
Comment (dignifiedquire) [inline src/test_xor128.rs]: this should be `read_exact_at`

File: src/test_xor128.rs [hunk 7]
```diff
@@ -465,6 +586,44 @@ fn test_various_trees_with_partial_cache() {
...
+                .read(true)
+                .open(StoreConfig::data_path(&config.path, &config.id))
+                .context("Failed to open base layer data")
+                .unwrap();
+            let mut base_layer = vec![0; count * 16];
+            reader.read(&mut base_layer)
+                .context("Failed to read")
+                .unwrap();
+
+            let output_file = temp_path.join(format!("base-data-only-{}", i));
```
Comment (dignifiedquire) [inline src/test_xor128.rs]: can use `std::fs::write`

File: src/test_xor128.rs [hunk 8]
```diff
@@ -475,10 +634,20 @@ fn test_various_trees_with_partial_cache() {
...
             // Then re-create an MT using LevelCacheStore and generate all proofs.
-            let level_cache_store: LevelCacheStore<[u8; 16]> =
+            let mut level_cache_store: LevelCacheStore<[u8; 16], _> =
                 Store::new_from_disk(2 * count - 1, &config).unwrap();
-            let mt_level_cache: MerkleTree<[u8; 16], XOR128, LevelCacheStore<_>> =
-                MerkleTree::from_data_store(level_cache_store, count).unwrap();
+            level_cache_store.set_external_reader(ExternalReader {
+                source: reader,
+                read_fn: |start, end, buf, reader| {
+                    reader.read_at(&mut buf[0..end - start], start as u64)
```
Comment (dignifiedquire) [inline src/test_xor128.rs]: `read_exact_at`

**Comment by cryptonemo** [comment]
Rebased against current master.

File: src/store.rs [hunk 9]
```diff
@@ -610,18 +670,98 @@ pub struct LevelCacheStore<E: Element> {
...
+            .field("elem_len", &self.len)
+            .field("data_width", &self.data_width)
+            .field("cache_index_start", &self.cache_index_start)
+            .field("store_size", &self.store_size)
+            .finish()
+    }
+}
+
+impl<E: Element, R: Read + Send + Sync> LevelCacheStore<E, R> {
+    // Used for opening v2 compacted DiskStores.
```
Comment (dignifiedquire) [inline src/store.rs]: please use doc comments fo these: `/// `

File: src/store.rs [hunk 10]
```diff
@@ -610,18 +670,98 @@ pub struct LevelCacheStore<E: Element> {
...
+    pub fn new_from_disk_with_reader(
+        store_range: usize,
+        config: &StoreConfig,
+        reader: Option<ExternalReader<R>>,
+    ) -> Result<Self> {
```
Comment (dignifiedquire) [inline src/store.rs]: I am not seeing any check that the config actually points to a `v2` store

File: src/store.rs [hunk 11]
```diff
@@ -610,18 +670,98 @@ pub struct LevelCacheStore<E: Element> {
...
+            .finish()
+    }
+}
+
+impl<E: Element, R: Read + Send + Sync> LevelCacheStore<E, R> {
+    // Used for opening v2 compacted DiskStores.
+    pub fn new_from_disk_with_reader(
+        store_range: usize,
+        config: &StoreConfig,
+        reader: Option<ExternalReader<R>>,
```
Comment (dignifiedquire) [inline src/store.rs]: this should not be an option, as the point of this method is to take an `ExternalReader`

File: src/store.rs [hunk 12]
```diff
@@ -824,10 +963,32 @@ impl<E: Element> LevelCacheStore<E> {
             "out of bounds"
         );
 
+        // If an external reader was specified for the base layer, use it.
+        if start < self.data_width * self.elem_len && self.reader.is_some() {
+            self.reader
+                .as_ref()
+                .unwrap()
+                .read(start, end, &mut read_data)
```
Comment (dignifiedquire) [inline src/store.rs]: should be `read_exact`

File: src/store.rs [hunk 13]
```diff
@@ -843,15 +1004,45 @@ impl<E: Element> LevelCacheStore<E> {
...
-                )
-            })?;
+        assert!(start <= self.data_width * self.elem_len || start >= self.cache_index_start);
+
+        // If an external reader was specified for the base layer, use it.
+        if start < self.data_width * self.elem_len && self.reader.is_some() {
+            self.reader
+                .as_ref()
+                .unwrap()
+                .read(start, end, buf)
```
Comment (dignifiedquire) [inline src/store.rs]: should be `read_exact`

File: src/store.rs (hunk 10 above)
Comment (cryptonemo) [reply src/store.rs]: The v2 check is the store size check: `store_size == cache_size`

**Pull request merged**